### PR TITLE
Add Firebase authentication, Blobstore image storage, and Firebase Firestore read/writes.

### PR DESCRIPTION
**Pull request by saraholiviawu**
**Authentication description**: Added backend to authentication pages signup.html, signupmethod.html, and login.html that registers a user for the existing browser session using the Firebase Authentication API with a Google Sign-in pop up. When the user first signs up, their display name, user id, email, and account type is stored in Firebase Firestore. An email verification is sent to their email address. When an existing user signs in, no data will be written to Firebase Firestore. The user will be logged in until he / she selects 'Log out' on the nav bar drop down menu.

**Blobstore  general description**: Images are served using the Blobstore API with Blobstore. 

**Account settings description**: Completed frontend and backend for accountsettings.html page. The page allows a user to add change their avatar photo. If the user has not yet changed their photo, the default avatar will be displayed. Furthermore, the page displays user account information about name, email, account type, and address by querying Firebase Firestore and placing the data onto HTML dynamically using the DOM. Note that the navbar avatar responds to the image changes.

**Blobstore on accountsettings.html**: When the user uploads a new blob, the blob-key is passed as a parameter to the same page. On the onload of the body, the blob key is parsed from the parameter and stored in Firebase. The image is then served directly as a blob. When no new blob-key is added, a null blob key is parsed from the parameter. The existing blob key is read from Firebase and the image is served as a blob. 

**Account settings demo**:
account_settings

### CODE REVIEW

File: ripple/target/appengine-staging/__static__/auth.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+function addNewUser(uid, name, isBusinessOwner, email) {
+  console.log("addNewUser("+ uid + ", " + name + ", " + isBusinessOwner + ")");
+  db.collection("users").doc(uid).set({
+    uid: uid,
+    name: name,
+    isBusinessOwner: isBusinessOwner,
+    email: email,
+    // blobKey for default avatar photo
+    blobKey: "XEkBWMfbItUDt70PromXrQ",
```
Comment (sabadas) [inline ripple/target/appengine-staging/__static__/auth.js:37]: Now that I am seeing it again, is this something that is provided and you are just adding it to github?

File: ripple/target/appengine-staging/__static__/auth.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+// function to sign in with Google
+function signInWithGoogle() {
+  // Signed in state will persist until user logs out 
+  firebase.auth().setPersistence(firebase.auth.Auth.Persistence.LOCAL)
+    .then(function() {
+      var provider = new firebase.auth.GoogleAuthProvider();
+      auth.signInWithPopup(provider).then(function(result) {
+        // window.location = 'landingbusiness.html';
```
Comment (sabadas) [inline ripple/target/appengine-staging/__static__/auth.js:54]: nit:remove

**Comment by sabadas** [comment]
This is a rather large CL and (correct me if I am wrong) I am assuming a lot of this was adding specific files to set up the Firebase authentication. Do all these files need a thorough review?

The GIF you posted looks great though! Only comment I have (and doesn't need to be addressed in this PR, is that you can see the account image kind of resize before it loads the new one. Is it possible to give it some fixed or min dimensions so that it doesn't have that jank?

File: ripple/src/main/webapp/auth.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    sendEmailVerification(uid);
+    addNewUser(uid, name, isBusinessOwner, email);
+  }).catch(function(error) {
+    console.log(error);
+    // Handle Errors here.
+    var errorCode = error.code;
+    var errorMessage = error.message;
+    var email = error.email;
+    var credential = error.credential;
+    // ...
```
[the remark this answers is not in the document]
Comment (saraholiviawu) [reply ripple/src/main/webapp/auth.js:28]: Fixed

File: ripple/src/main/webapp/auth.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+function addNewUser(uid, name, isBusinessOwner, email) {
+  console.log("addNewUser("+ uid + ", " + name + ", " + isBusinessOwner + "," + email + ")");
+  db.collection("users").doc(uid).set({
+    uid: uid,
+    name: name,
+    isBusinessOwner: isBusinessOwner,
+    email: email,
+    address: "555 Fake Ln, Apt 102",
+    // blobKey for default avatar photo
+    blobKey: "XEkBWMfbItUDt70PromXrQ",
```
[the remark this answers is not in the document]
Comment (saraholiviawu) [reply ripple/src/main/webapp/auth.js:42]: This is a blob that I generated. Yes, I can create a resource file to refer to variables like this!

File: ripple/target/appengine-staging/__static__/auth.js (hunk 1 above)
Comment (saraholiviawu) [reply ripple/target/appengine-staging/__static__/auth.js:37]: I believe target directory files are automatically generated based on the code stored in my src directory. Changes in the src are detected by the target and changed accordingly.

File: ripple/target/appengine-staging/__static__/auth.js (hunk 2 above)
Comment (saraholiviawu) [reply ripple/target/appengine-staging/__static__/auth.js:54]: I have uncommented the line! Fixed.

Comment (saraholiviawu) [reply ripple/target/appengine-staging/__static__/auth.js:54]: In my src auth.js file, the line is uncommented. I believe this target file should automatically update on its own.

**Comment by saraholiviawu** [comment]
Many of these files do have to do with Firebase config. Only files in the ripple/src files need thorough review. 

In a future pull request, I will give the account info a fixed or min dimensions to behavior that avoid the loading behavior we see currently.

**Pull request merged**